core/extensions/impl/storage_extension.cpp

```diff
@@ -3,99 +3,188 @@
-    std::terminate();
+  runtime::WasmPointer StorageExtension::ext_get_allocated_storage(
+      runtime::WasmPointer key_data, runtime::SizeType key_length,
+      runtime::WasmPointer len_ptr) {
+    auto key = memory_->loadN(key_data, key_length);
+    auto data = db_->get(key);
+    const auto length = data.has_value() ? data.value().size()
+                                         : runtime::WasmMemory::kMaxMemorySize;
+    memory_->store32(len_ptr, length);
+
+    if (not data) {
+      return 0;
+    }
+
+    auto data_ptr = memory_->allocate(length);
+
+    if (data_ptr != -1) {
+      memory_->storeBuffer(data_ptr, data.value());
+    }
+    return data_ptr;
   }
 
-  uint32_t StorageExtension::ext_exists_child_storage(
-      const uint8_t *storage_key_data, uint32_t storage_key_length,
-      const uint8_t *key_data, uint32_t key_length) {
-    std::terminate();
+  runtime::SizeType StorageExtension::ext_get_storage_into(
+      runtime::WasmPointer key_data, runtime::SizeType key_length,
+      runtime::WasmPointer value_data, runtime::SizeType value_length,
+      runtime::SizeType value_offset) {
+    auto key = memory_->loadN(key_data, key_length);
+    auto data = get(key, value_offset, value_length);
+    if (not data) {
+      return runtime::WasmMemory::kMaxMemorySize;
+    }
+    memory_->storeBuffer(value_data, *data);
+    return data->size();
   }
 
-  uint8_t *StorageExtension::ext_get_allocated_child_storage(
-      const uint8_t *storage_key_data, uint32_t storage_key_length,
-      const uint8_t *key_data, uint32_t key_length, uint32_t *written) {
-    std::terminate();
+  void StorageExtension::ext_set_storage(const runtime::WasmPointer key_data,
+                                         runtime::SizeType key_length,
+                                         const runtime::WasmPointer value_data,
+                                         runtime::SizeType value_length) {
+    auto key = memory_->loadN(key_data, key_length);
+    auto value = memory_->loadN(value_data, value_length);
+    auto put_result = db_->put(key, value);
+    if (not put_result) {
+      logger_->warn(
+          "ext_set_storage failed, due to fail in trie db with reason: {}",
+          put_result.error().message());
+    }
   }
 
-  uint8_t *StorageExtension::ext_get_allocated_storage(const uint8_t *key_data,
-                                                       uint32_t key_length,
-                                                       uint32_t *written) {
-    std::terminate();
+  // -------------------------Trie operations--------------------------
+
+  void StorageExtension::ext_blake2_256_enumerated_trie_root(
+      runtime::WasmPointer values_data, runtime::WasmPointer lens_data,
+      runtime::SizeType lens_length, runtime::WasmPointer result) {
+    std::vector<common::Buffer> data;
+    for (runtime::SizeType i = 0; i < lens_length; i++) {
+      auto length = memory_->load32s(lens_data + i * 4);
```

Comment:
The `length` should be crated outside of a loop in order to avoid creations in each cycle